File: app/src/main/java/pl/karol202/paintplus/image/layer/Layer.java
/*
 * Copyright 2017 karol-202
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package pl.karol202.paintplus.image.layer;

import android.graphics.*;
import pl.karol202.paintplus.image.Image.FlipDirection;
import pl.karol202.paintplus.image.Image.OnImageChangeListener;
import pl.karol202.paintplus.image.layer.mode.DefaultLayerMode;

public class Layer
{
	private OnImageChangeListener listener;
	private Bitmap bitmap;
	private Canvas editCanvas;

	private String name;
	private boolean visible;
	private boolean temporaryHidden;
	private int x;
	private int y;
	private LegacyLayerMode mode;
	private float opacity;

	public Layer(int x, int y, String name, int width, int height, int color)
	{
		this(x, y, name, Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888));
		bitmap.eraseColor(color);
	}

	public Layer(int x, int y, String name, Bitmap bitmap)
	{
		this.bitmap = bitmap;
		this.editCanvas = new Canvas(bitmap);

		this.name = name;
		this.visible = true;
		this.temporaryHidden = false;
		this.x = x;
		this.y = y;
		this.mode = new DefaultLayerMode(this);
		this.opacity = 1f;
	}

	public void offset(int x, int y)
	{
		this.x += x;
		this.y += y;
	}

	public void resize(int x, int y, int width, int height)
	{
		Bitmap source = bitmap;
		bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
		editCanvas = new Canvas(bitmap);
		editCanvas.drawBitmap(source, -x, -y, null);
		this.x += x;
		this.y += y;
	}

	public void scale(double scaleX, double scaleY, boolean bilinear)
	{
		int width = (int) Math.round(bitmap.getWidth() * scaleX);
		int height = (int) Math.round(bitmap.getHeight() * scaleY);
		scale(width, height, bilinear);
	}

	public void scale(int width, int height, boolean bilinear)
	{
		Matrix matrix = new Matrix();
		matrix.preScale((float) width / bitmap.getWidth(), (float) height / bitmap.getHeight());

		bitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), matrix, bilinear);
		editCanvas = new Canvas(bitmap);
	}

	public void flip(FlipDirection direction)
	{
		Matrix matrix = new Matrix();
		matrix.preScale(direction == FlipDirection.HORIZONTALLY ? -1 : 1, direction == FlipDirection.VERTICALLY ? -1 : 1);

		bitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), matrix, false);
		editCanvas = new Canvas(bitmap);
	}

	public void rotate(float angle)
	{
		rotate(angle, true);
	}

	public void rotate(float angle, boolean offset)
	{
		int oldWidth = bitmap.getWidth();
		int oldHeight = bitmap.getHeight();
		Matrix matrix = new Matrix();
		matrix.preRotate(angle);

		bitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), matrix, true);
		editCanvas = new Canvas(bitmap);

		if(!offset) return;
		x -= (bitmap.getWidth() - oldWidth) / 2;
		y -= (bitmap.getHeight() - oldHeight) / 2;
	}

	public Bitmap drawLayerAndReturnBitmap(Bitmap bitmap, Canvas canvas, RectF clipRect, Matrix imageMatrix)
	{
		Matrix layerMatrix = new Matrix(imageMatrix);
		layerMatrix.preTranslate(x, y);

		mode.startDrawing(bitmap, canvas);
		if(clipRect != null) mode.setRectClipping(clipRect);
		mode.addLayer(layerMatrix);
		if(clipRect != null) mode.resetClipping();
		return mode.apply();
	}

	public void setImageChangeListener(OnImageChangeListener listener)
	{
		this.listener = listener;
	}

	public Bitmap getBitmap()
	{
		return bitmap;
	}

	public void setBitmap(Bitmap bitmap)
	{
		if(!bitmap.isMutable()) bitmap = bitmap.copy(Bitmap.Config.ARGB_8888, true);
		bitmap.setHasAlpha(true);
		this.bitmap = bitmap;
		editCanvas = new Canvas(bitmap);
	}

	public Canvas getEditCanvas()
	{
		return editCanvas;
	}

	public int getWidth()
	{
		return bitmap.getWidth();
	}

	public int getHeight()
	{
		return bitmap.getHeight();
	}

	public Rect getBounds()
	{
		return new Rect(x, y, x + getWidth(), y + getHeight());
	}

	public String getName()
	{
		return name;
	}

	public void setName(String name)
	{
		this.name = name;
	}

	public boolean isVisible()
	{
		return visible;
	}

	public void setVisibility(boolean visible)
	{
		this.visible = visible;
		if(listener != null) listener.onImageChanged();
	}

	public boolean isTemporaryHidden()
	{
		return temporaryHidden;
	}

	public void setTemporaryHidden(boolean temporaryHidden)
	{
		this.temporaryHidden = temporaryHidden;
	}

	public int getX()
	{
		return x;
	}

	public int getY()
	{
		return y;
	}

	public void setPosition(int x, int y)
	{
		this.x = x;
		this.y = y;
		if(listener != null) listener.onImageChanged();
	}

	public LegacyLayerMode getMode()
	{
		return mode;
	}

	public void setMode(LegacyLayerMode mode)
	{
		this.mode = mode;
		mode.setLayer(this);
	}

	public float getOpacity()
	{
		return opacity;
	}

	public void setOpacity(float opacity)
	{
		this.opacity = opacity;
	}
}

File: app/src/main/java/pl/karol202/paintplus/image/layer/Layer.kt
/*
 * Copyright 2017 karol-202
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package pl.karol202.paintplus.image.layer

import android.graphics.*
import androidx.core.graphics.plus
import pl.karol202.paintplus.image.FlipDirection
import pl.karol202.paintplus.image.layer.mode.DefaultLayerMode
import pl.karol202.paintplus.image.layer.mode.LayerMode
import pl.karol202.paintplus.util.*
import java.util.*
import kotlin.math.roundToInt

class Layer private constructor(val id: String,
                                val x: Int,
                                val y: Int,
                                val name: String,
                                val bitmap: Bitmap,
                                val mode: LayerMode = DefaultLayerMode,
                                val visible: Boolean = true,
                                val opacity: Float = 1f)
{
	companion object
	{
		fun create(x: Int = 0, y: Int = 0, name: String, bitmap: Bitmap) =
				Layer(createRandomId(), x, y, name, bitmap.ensureMutable().withAlpha(true))

		fun create(x: Int = 0, y: Int = 0, name: String, width: Int, height: Int, color: Int) =
				create(x, y, name, Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)).apply {
					bitmap.eraseColor(color)
				}

		private fun createRandomId() = UUID.randomUUID().toString()
	}

	val editCanvas = Canvas(bitmap)

	val topLeft = Point(x, y)
	val width = bitmap.width
	val height = bitmap.height
	val bounds = Rect(x, y, x + width, y + height)
	val matrix = Matrix().preTranslated(x.toFloat(), y.toFloat())

	fun translated(x: Int, y: Int) =
			withPosition(this.x + x, this.y + y)

	fun resized(x: Int, y: Int, width: Int, height: Int) =
			withEmptyBitmap(width, height).translated(x - this.x, y - this.y).apply {
				editCanvas.drawBitmap(this@Layer.bitmap, (this@Layer.x - x).toFloat(), (this@Layer.y - y).toFloat(), null)
			}

	fun scaled(width: Int, height: Int, bilinear: Boolean) =
			withBitmap(bitmap.transformedWith(Matrix().preScaled(width.toFloat() / bitmap.width,
			                                                     height.toFloat() / bitmap.height), bilinear))

	fun scaled(scaleX: Float, scaleY: Float, bilinear: Boolean) =
			scaled((bitmap.width * scaleX).roundToInt(), (bitmap.height * scaleY).roundToInt(), bilinear)

	fun flipped(direction: FlipDirection) =
			scaled(if(direction == FlipDirection.HORIZONTALLY) -1f else 1f,
			       if(direction == FlipDirection.VERTICALLY) -1f else 1f, bilinear = false)

	fun rotated(angle: Float, offset: Boolean = true): Layer
	{
		val newBitmap = bitmap.transformedWith(Matrix().preRotated(angle), bilinear = true)
		return withBitmap(newBitmap).translated(if(offset) (newBitmap.width - width) / 2 else 0,
		                                        if(offset) (newBitmap.height - height) / 2 else 0)
	}

	fun duplicated() = copy(id = createRandomId(), bitmap = bitmap.duplicated())

	fun withPosition(x: Int, y: Int) = copy(x = x, y = y)

	fun withName(name: String) = copy(name = name)

	fun withBitmap(bitmap: Bitmap) =
			copy(bitmap = bitmap.ensureMutable().withAlpha(true))

	private fun withEmptyBitmap(width: Int, height: Int) =
			withBitmap(Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888))

	fun withMode(mode: LayerMode) = copy(mode = mode)

	fun withVisibility(visible: Boolean) = copy(visible = visible)

	fun withOpacity(opacity: Float) = copy(opacity = opacity)

	private fun copy(id: String = this.id, x: Int = this.x, y: Int = this.x, name: String = this.name,
	                 bitmap: Bitmap = this.bitmap, mode: LayerMode = this.mode, visible: Boolean = this.visible,
	                 opacity: Float = this.opacity) =
			Layer(id, x, y, name, bitmap, mode, visible, opacity)
}
